module5/lessons/Week4/ProcessOptimization.md

```diff
@@ -0,0 +1,246 @@
+    * What if num[1] is a duplicate of num[0]?
+    * What if num[9] is a duplicate of num[5]?
+* How many steps would there be if `nums` had 25 elements?
+* How many steps would there be if `nums` had 50 elements?
+
+**Option 1**  
+```c#
+// find the duplicate
+int[] nums = {36, 84, 75, 20, 30, 44, 17, 98, 92, 27 };
+int? duplicate;
+
+foreach (var numA in nums)
+{
+    foreach (var numB in nums)
+    {
+        if (numA == numB)
+        {
+            duplicate = num1;
+        }
+    }
+}
+
+Console.WriteLine(duplicate);
+```
+**Option 2**
+```c#
+int[] nums = { 36, 84, 75, 20, 92, 44, 17, 98, 92, 27 };
+Dictionary<int, int> countOfNums = new Dictionary<int, int>();
+
+foreach (var num in nums)
+{
+    if (countOfNums.ContainsKey(num))
+    {
+        Console.WriteLine(num);
+        break;
+    }
+    countOfNums.Add(num, 1);
+}
+```
+</section>
+
+## Big O
+
+It's not very useful to have to describe algorithms in *exact* number of steps for every possible number of elements.  So, developers describe time complexity using mathematic expressions.  We call these expressions **Big O Notation** (big 'oh').  The BigO of an algorithm describes the _worst-case scenario_: what is the maximum number of stes an algorithm will take in relation to how the dataset grows.
+
+When describing BigO, we `"n"` to signify the amount of data. The most common examples you might see include:
+
+- O(1)
+- O(log n)
+- O(n)
+- O(n log n)
+- O(n^2) (aka, n-squared)
+
+We'll discuss each of these in order from least complex (best performance) to most complex (worst performance).
+
+<section class="answer" markdown='1'>
+### Constant Time - O(1)
+
+**Description:**
+- No matter how much data there is, it only takes one operation.
+
+**Examples:**
+- Simple math operations and assignment operators
+- Invoking a function
+- Accessing an array by index position
+- Accessing a property in an object/hash by its attribute/key
+</section>
+
+<section class="answer" markdown='1'>
+### Logarithmic Time - O(log n)
+
+**Description:**
+- Every step/action you take, you cut out half of the work.
+
+**Examples:**
+- "Binary search" using a binary tree.  An example of doing this might include looking up a word in a dictionary.
+</section>
+
+<section class="answer" markdown='1'>
+### Linear Time - O(n)
+
+**Description:**
+- Complexity grows in direct proportion to the size of the input data.
+
+**Examples:**
+- Iteration over an array or the keys of an object/hash with a `for` loop or `while` loop
+- Using array prototype methods/enumerables such as `map` or `reduce`
+- Anything that looks at, or processes, every element in a data structure.
+</section>
+
+<section class="answer" markdown='1'>
+### Linearithmic Time - O(n log n)
+
+**Description:**
+- Often is a nested iteration where within each operation done in linear time, there are actions being done in logarithmic time over the same size of data
+
+**Examples:**
+- Efficient sorting algorithms such as `merge sort`
```

Comment:
It's an interesting question if we want to teach searching/sorting algorithms. I think we're leaning towards no, and we will pivot if we learn students are being asked those types of questions.

In that case I would not assume students will learn what merge sort is, this makes it sort of seem like a thing they should know. I would make this line stronger and say something like "The most efficient algorithms for sorting a list (quicksort and mergesort are the names of two very efficient sorting algorithms!)"